Source/Core/InputCommon/ControllerInterface/CemuHookUDPServer/CemuHookUDPServer.cpp

```diff
@@ -0,0 +1,421 @@
+    {Config::System::CemuHookUdpServer, "Server", "IPAddress"}, DEFAULT_SERVER_ADDRESS};
+const Config::ConfigInfo<int> SERVER_PORT{{Config::System::CemuHookUdpServer, "Server", "Port"},
+                                          DEFAULT_SERVER_PORT};
+}  // namespace Settings
+
+static bool s_server_enabled;
+static std::string s_server_address;
+static u16 s_server_port;
+static u32 s_client_uid;
+static std::chrono::steady_clock::time_point s_next_listports;
+static std::thread s_hotplug_thread;
+static Common::Flag s_hotplug_thread_running;
+std::mutex s_port_info_mutex;
+Proto::MessageType::PortInfo s_port_info[Proto::PORT_COUNT];
+
+static bool IsSameController(const Proto::MessageType::PortInfo& a,
+                             const Proto::MessageType::PortInfo& b)
+{
+  // compare everything but battery_status
+  return a.pad_id == b.pad_id && a.pad_state == b.pad_state && a.model == b.model &&
+         a.connection_type == b.connection_type &&
+         memcmp(a.pad_mac_address, b.pad_mac_address, sizeof a.pad_mac_address) == 0;
+}
+
+static sf::Socket::Status ReceiveWithTimeout(sf::UdpSocket& socket, void* data, std::size_t size,
+                                             std::size_t& received, sf::IpAddress& remoteAddress,
+                                             unsigned short& remotePort, sf::Time timeout)
+{
+  sf::SocketSelector selector;
+  selector.add(socket);
+  if (selector.wait(timeout))
+    return socket.receive(data, size, received, remoteAddress, remotePort);
+  else
+    return sf::Socket::NotReady;
+}
+
+static void HotplugThreadFunc()
+{
+  Common::SetCurrentThreadName("CemuHookUDPServer Hotplug Thread");
+  NOTICE_LOG(SERIALINTERFACE, "CemuHookUDPServer hotplug thread started");
+
+  sf::UdpSocket socket;
+
+  while (s_hotplug_thread_running.IsSet())
+  {
+    const auto now = std::chrono::steady_clock::now();
+    if (now >= s_next_listports)
+    {
+      s_next_listports = now + SERVER_LISTPORTS_INTERVAL;
+
+      // Request info on the four controller ports
+      Proto::Message<Proto::MessageType::ListPorts> list_ports(s_client_uid);
+      list_ports.pad_request_count = 4;
+      list_ports.pad_id[0] = 0;
+      list_ports.pad_id[1] = 1;
+      list_ports.pad_id[2] = 2;
+      list_ports.pad_id[3] = 3;
+      list_ports.Finish();
+      if (socket.send(&list_ports, sizeof list_ports, s_server_address, s_server_port) !=
+          sf::Socket::Status::Done)
+        ERROR_LOG(SERIALINTERFACE, "CemuHookUDPServer HotplugThreadFunc send failed");
+    }
+
+    // Receive controller port info
+    Proto::Message<Proto::MessageType::FromServer> msg;
+    const auto timeout = s_next_listports - std::chrono::steady_clock::now();
+    // ReceiveWithTimeout treats a timeout of zero as infinite timeout, which we don't want
+    auto timeout_ms = std::chrono::duration_cast<std::chrono::milliseconds>(timeout).count();
+    timeout_ms = std::max(timeout_ms, 1LL);
```

Comment:
This fails to compile here in Linux where `long long` is not the same as `std::chrono::milliseconds::rep`.
`timeout_ms = std::max<decltype(timeout_ms)>(timeout_ms, 1);` works.